src/Npgsql/PostgresException.cs

```diff
@@ -293,19 +293,25 @@ public override IDictionary Data
         {
             get
             {
-                // Remarks: return Dictionary with object keys although all our keys are string keys
-                // because System.Windows.Threading.Dispatcher relies on that
-                return _data ?? (_data = (
+                if (_dataInitialized)
+                    return base.Data;
+
+                var data = base.Data;
```

Comment:
Just out of curiosity, does the (release/optimized) IL actually look different when accessing the base member directly in the loop as opposed to stashing in a local?

Reply:
It isn't a responsibility of the compiler but the JIT. The base method is located in a different assembly and you don't know how exactly it's implemented until runtime. As I know the JIT doesn't perform such optimizations because it isn't a simple field access under the hood, but will check.

Reply:
Yes, the JIT doesn't optimize away base property access in that case.

Reply:
That's disappointing, but makes sense... Thanks for checking for me!

Reply:
Why so? When you call a method multiple times they must not be eliminated since its result can differ from time to time. Therefore, caching a property value improves performance when you're accessing it multiple times like in my case.

Reply:
> Why so? When you call a method multiple times they must not be eliminated since its result can differ from time to time.

Sure. But for a base class property accessor, I was wishfully thinking about a const-function-like optimization, where the JITer might cache into a local on it's own.

Reply:
I know that there has been a lot of work on devirtualization and similar optimizations in the JIT. I would be surprised if methods calls on base would be optimized (if not now, maybe later). This isn't the same as caching, which does seem possible in this specific case since `this` cannot change...

What's really annoying is that [sharplab.io](https://sharplab.io) still uses .NET Framework rather than .NET Core - it would have been great to easily see what asm the latest JIT produces...

(not that any of this matters much here :))

Reply:
`this` cannot change, but its state can. There is no proof for the compiler or the JIT that the state is constant without a deep and expensive code analysis. So the only one way to reduce calls is to store the value in a local variable.

Reply:
Right, I was referring to this itself, but you're of course right that any property or method result cannot be cached.